### DISCUSSION TITLE
Converting Python testscript to TESTed - errors with OOP excercise

### DISCUSSION THREAD

**Discussion by marcothijs** [dodona-edu/feedback-and-support#30] [Questions - Vragen] [answered]
Dear, 

In my attempt to move the testscript to TESTed, I run into error I can't seem to fix. I have the following exercise: https://dodona.be/nl/activities/912288875/ with a testscript using "pythia_judge": "doctest" with all the tests I want to execute which work well. I have made a copy of this exercise: https://dodona.be/nl/activities/844895513/ with an attempt to test is with TESTed, but run into a couple of issues:
- can't test __repr__ with print(), worked-a-round the error by using the __repr__ methode directly (is ok as solution but why can't I use print()
- can't test __eq__ with the a comparison + print(), worked-a-round the error by using the __eq__ methode directly (is ok as solution but same question as above)
- can't change an attribute with "- statement: 'p_h1.y = 3.0' ", get an error and have no idea how to get this right... (put in comment for now in the testscript). I'm using this exact code to test the copy() function and don't see a work-a-round.

Further, I was able to get the assertion error tested.

Thanks in advance for your help.

Rgds,
Marco

**Comment by bmesuere** [comment]
Hi @marcothijs 

I noticed this question when going through open notifications. We probably missed this one during the Christmas break. Sorry!
Are you still having this problem?

**Reply by marcothijs** [reply]
Hi @bmesuere,

I still have the issue as I was waiting for a reply from Dodona, but I wasn't in a hurry as I still had the pythia_judge way to test. I think it is best to look an my excercise (see links above) to get the picture of my questions/issues. 

Thanks for looking into it. If my explanation wasn't clear, please let me know. 

PS. I would be nice to have all the examples available for the pythia_judge translated to TESTed scripts. That way I might be able to figure out on my own how to translate my own pythia_judge scripts into TESTed  scripts. I seems to have trouble to do it based on the available examples. 

Thanks in advance. 
Marco

**Comment by niknetniko** [comment, marked as answer]
Also my apologies for missing this!

----

For the first two points, the issue is that we currently don't support running arbitrary functions, and don't have a programming-language independent way of doing print/repr (and there are no concrete plans for supporting it soon).

However, we do have language-specific literals for this use case (see the docs [here](https://docs.dodona.be/nl/guides/exercises/testsuites/#taalspecifieke-expressies-en-statements)). They do have some drawbacks, but they are a good fit in this case.

For example, transform this:

```yaml
- statement: 'punt1 = Punt(1.0, 1.0)'
- expression: 'punt1.__repr__()'
  return: "(1.0, 1.0)"
```

To this:

```yaml
- statement: 'punt1 = Punt(1.0, 1.0)'
- expression:
    python: 'repr(punt1)'
  return: "(1.0, 1.0)"
```

This way, the expression is not limited to the language-independent DSL of TESTed, but you use arbitrary Python code.

----

The assignment of properties is also not supported, but that is more of an oversight, we do want to do this at some point (see https://github.com/dodona-edu/universal-judge/issues/500).
Until it is implemented, you could also use language-specific literals instead:

```yaml
- statement:
    python: 'p_h1.y = 3.0'
- expression: 'p_h1.y'
  return: 3.0
```

**Reply by marcothijs** [reply]
Thanks @bmesuere, @niknetniko for your help!
